src/indieweb_utils/indieauth/flask.py

```diff
@@ -23,15 +41,17 @@ def indieauth_callback_handler(
     :type callback_url: str
     :param client_id: The client ID used in the original authentication request.
     :type client_id: str
-    :param required_scopes: The scopes required for the application to work. This list should not include optional scopes.
+    :param required_scopes: The scopes required for the application to work.
+            This list should not include optional scopes.
     :type required_scopes: list[str]
-    :return: A message indicating the result of the callback (success or failure) and the token endpoint response. The endpoint response will be equal to None if the callback failed.
+    :return: A message indicating the result of the callback (success or failure) and the token endpoint response.
+            The endpoint response will be equal to None if the callback failed.
     :rtype: tuple[str, dict]
     """
 
     if state != session_state:
-        message = "Your authentication failed. Please try again."
-        return message, None
+        message = "The provided state value did not match the session state. Please try again."
+        raise AuthenticationError(message)
 
     data = {
```

Comment:
I wonder if the "making a request" part of this couldn't be broken into a smaller function that could be re-used.